LandLord-latest/src/main/java/biz/princeps/landlord/manager/WorldGuardManager.java

```diff
@@ -280,4 +287,74 @@ public boolean isAllowedInOverlap(Player p, Location loc) {
         return false;
     }
 
+    @Override
+    public Pair<Integer, Integer> calcClaimHeightBoundaries(Chunk chunk) {
+        ClaimHeightDefinition boundaryMethod = ClaimHeightDefinition.parse(plugin.getConfig().getString("ClaimHeight.method"));
+
+        // We will use the full and default behaviour as default value.
+        if (boundaryMethod == null) {
+            boundaryMethod = ClaimHeightDefinition.FULL;
+        }
+
+        World world = chunk.getWorld();
+        com.sk89q.worldedit.world.World weWorld = BukkitAdapter.adapt(world);
+        int maxHeight = weWorld.getMaxY();
+        int minHeight = weWorld.getMinY();
+
+        // Full is the default behaviour.
+        // This will claim the whole chunk.
+        if (boundaryMethod == ClaimHeightDefinition.FULL) {
+            return Pair.of(minHeight, maxHeight);
+        }
+
+        String claimHeightWorldType = world.getEnvironment() == World.Environment.NORMAL ? "overworld" : "other";
+        int bottomY = plugin.getConfig().getInt("ClaimHeight." + claimHeightWorldType + "-bottomY", minHeight);
+        int topY = plugin.getConfig().getInt("ClaimHeight." + claimHeightWorldType + "-topY", maxHeight);
+
+        // Fixed is the simple claim behaviour.
+        // We want to handle this first.
+        if (boundaryMethod == ClaimHeightDefinition.FIXED) {
+            return Pair.of(Math.max(minHeight, bottomY), Math.min(topY, maxHeight));
+        }
+
+        // Let's find all highest points in the chunk.
+        List<Integer> points = new ArrayList<>();
+        for (int x = 0; x < 16; x++) {
+            for (int z = 0; z < 16; z++) {
+                points.add(world.getHighestBlockYAt((chunk.getX() << 4) + x, (chunk.getZ() << 4) + z));
+            }
+        }
+
+        // Get the center based on the boundary Method
+        int center = boundaryMethod.getCenter(points);
+
+        bottomY = center + bottomY;
+        topY = center + topY;
+
+        if (plugin.getConfig().getBoolean("ClaimHeight.appendOversize")) {
+            // We append the oversize which reach out of the world on the top or the bottom if it fits.
+            // We throw oversize away if we would exceed the world height limit on both ends.
+            if (topY > maxHeight) {
+                bottomY -= topY - maxHeight;
+                topY = maxHeight;
+                if (bottomY < minHeight) {
+                    bottomY = minHeight;
+                }
+            }
+
+            if (bottomY < minHeight) {
+                topY += Math.abs(bottomY);
+                bottomY = minHeight;
+                if (topY > maxHeight) {
+                    topY = maxHeight;
+                }
+            }
+        } else {
+            // Just clamp this stuff.
+            bottomY = Math.max(bottomY, minHeight);
+            topY = Math.min(topY, maxHeight);
+        }
+        return Pair.of(bottomY, topY);
+    }
```

Comment:
This is a lot of duplicated code with the legacy module.
Maybe add a overload with calcClaimHeightBoundaries(Chunk chunk, int minHeigh, int maxHeigh) and extract the common logic.